src/Runners/PHPUnit/Options.php

```diff
@@ -113,6 +113,10 @@ public function __construct(array $opts = [])
         foreach (self::defaults() as $opt => $value) {
             $opts[$opt] = $opts[$opt] ?? $value;
         }
+        
+        if ($opts['processes'] === 'auto') {
+            $opts['processes'] = $this->getNumberOfCPUCores();
```

Comment:
I'm not sure but I think there should be at least 2 processes since that helps even on a single core machine and some build pipelines have just a single core assigned.

Reply:
I was also thinking about the best value when reviewing the PR. But we also want to avoid that the system starts lacking, etc. What value per core do you use?

Reply:
Personally I use cores - 1

So one is left for parent, also to not freeze the machine

Reply:
Machine freezing is more of an io issue on linux than cpu resources. There is `ionice` for that.

But I see that there are a lot of cases. I wouldn't overload hyperthreads while I'd definitly do that if I only had real cores. My idea of "at least 2" stems from build pipelines with 1 virtual core. Also It wouldn't make much sense to run paratest with 1 process. But I also admit that I could just define 2 processes in my build pipeline and for the local machine it's probably fine to have just as many processes as cores.

Since the default is now "auto" another feasible solution would be `100%` or `150%` to define over/underloading the machine.

BUT its probably best to leave it as is and merge this first and than think about a better algorithm if there is a need for it.

Reply:
We can make a special case when value is 1 and set it to 2.
Meaning two is minimum, this can make sense as it is also the default.

**Or** we ca change "auto" to introduce a set of strategies, either percentages as you mention or something ala:
- Cores
- Coresx2 ("hyper threading ceiling mode")
- Cores+1
- Cores-1

Where we pick on of them as default.
But I think I would prefer percentages to keep things "simple".

Either way this can make sense given there are as you mention many use cases.
If a given test suite is set of integration testing using things like cache service and maybe also full database then there are plenty of other processes already competing for available system resources.

Reply:
Not sure about this. I also thought yesterday about it and that it might be nice to set a dynamic value relative to the number of CPU cores. But it might already be too complicated and actually I think the setting to set the total number of processes is already absolutely sufficient and fine.

I mean, if someone has 4 cores and wants to use the strategy "cores x 2", he can just set it to 8. If someone has 10 cores and wants to use the strategy "cores + 1", he can just set it to 11 by himself. What I mean is, if someone sets the value manually anyway and changes the config, he already put thought into it. Having 2 settings for it then (like a strategy for the number of CPU cores and also the total number of processes) would just be a duplicate and imo, a useless complication.

So for me, I would just say that "auto" should provide a reasonable default. If you want to change it to something different, the developer should just set the number of processes accordingly and thus apply whatever strategy he likes.

And I also think "cores x 2" is a good default strategy - it shouldn't make the host lack, etc. Then, we can also skip the min value of 2, as every machine obviously has 1 core already, which leads to a min value of 2.

Reply:
I like the idea. Just a simple selection. If you are using `paratest` you probably want your tests to be done fast so just default to process spam. If it annoys you, then adjust it manually. My build pipelines have 1 core so 2 is what I was using anyways but if I'd port my app somewhere else with 2 vcores, 4 is fine.

I did some testing with a project i'm working on. I have 2 cores 4 threads. Running the functional tests with 8 threads isn't much different than running them with 5 but it is definitely faster than running them with 4 threads.

The only concern with that solution is how well paratest handles massive amounts of processes like on a Threadripper with 32 cores and 64 threads. It would then run with 128 processes.

Reply:
> The only concern with that solution is how well paratest handles massive amounts of processes like on a Threadripper with 32 cores and 64 threads. It would then run with 128 processes.

As far as I understand @andreasschroth it will be 64, cores (32) x2.  At least the intention for the code here is to count the physical cores and not the virtual once.

Reply:
@Nemo64: I believe (didn't test it myself though) that there shouldn't be any problems. And not sure if the code from @andrerom retrieves the number of physical cores or the logical ones. But normally for the OS there's no difference, as it only works with the logical cores usually. At least, that's what I know. So yes, I agree with you.

But setting up some DigitalOcean droplet for example for 1 hour and running some tests on 16 or 24 vCPUs shouldn't be a big deal to test it.

Reply:
So summary, what should be changed here?

The core count x2 when set to auto?

Reply:
@andrerom Thanks. I'm just missing the min value of 2, if the server just has 1 core... Otherwise paratest wouldn't serve any purpose.